addons/gamification/models/gamification_challenge.py

```diff
@@ -218,6 +218,9 @@ def write(self, vals):
 
         elif vals.get('state') == 'done':
             self._check_challenge_reward(force=True)
+            self.env['gamification.goal'].search([
+                ('challenge_id', 'in', self.ids),
+            ]).update_goal()
```

Comment:
Shouldn't we have this on the line before to update goals before the rewards allocation? Unless there is a reason not too, it feels more natural doesn't it?

Reply:
I probably had some idea behind this, but this indeed creates another problem with the rewards. I'll change it.

Reply:
I added a check in the unit test to make sure the user gets awarded with a badge.